src/systems/particle_emitter/ParticleEmitter.cc

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <ignition/msgs/particle_emitter.pb.h>
+
+#include <string>
+
+#include <ignition/common/Profiler.hh>
+#include <ignition/math/Color.hh>
+#include <ignition/math/Vector3.hh>
+#include <ignition/msgs/Utility.hh>
+#include <ignition/plugin/Register.hh>
+
+#include <ignition/gazebo/components/Name.hh>
+#include <ignition/gazebo/components/ParticleEmitter.hh>
+#include <ignition/gazebo/components/Pose.hh>
+#include <ignition/gazebo/Conversions.hh>
+#include <sdf/Material.hh>
+#include "ParticleEmitter.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+// Private data class.
+class ignition::gazebo::systems::ParticleEmitterPrivate
+{
+  /// \brief The particle emitter.
+  public: ignition::msgs::ParticleEmitter emitter;
+
+  /// \brief Get a RGBA color representation based on a color's
+  /// string representation.
+  /// \param[in] _colorStr The string representation of a color (ex: "black"),
+  /// which is case sensitive (the string representation should be lowercase).
+  /// \return The Color, represented in RGBA format. If _colorStr is invalid,
+  /// ignition::math::Color::White is returned
+  public: ignition::math::Color GetColor(const std::string &_colorStr) const;
+};
+
+//////////////////////////////////////////////////
+ParticleEmitter::ParticleEmitter()
+  : System(), dataPtr(std::make_unique<ParticleEmitterPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter::Configure(const Entity &/*_entity*/,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &/*_eventMgr*/)
+{
+  // Create a particle emitter entity.
+  auto entity = _ecm.CreateEntity();
+  if (entity == kNullEntity)
+  {
+    ignerr << "Failed to create a particle emitter entity" << std::endl;
+    return;
+  }
+
+  // Name.
+  std::string name = "particle_emitter_entity_" + std::to_string(entity);
+  if (_sdf->HasElement("emitter_name"))
+    name = _sdf->Get<std::string>("emitter_name");
+  this->dataPtr->emitter.set_name(name);
```

Comment:
Similar to my comment about a [potential redundant pose field](https://github.com/ignitionrobotics/ign-gazebo/pull/562/files#r572451528) in `emitter`, I also wonder if the name field is redundant for `emitter`. We store the name as its own component later in this method: https://github.com/ignitionrobotics/ign-gazebo/blob/9ed110b9105131a8c82ae86013bd8fdb4264fc21/src/systems/particle_emitter/ParticleEmitter.cc#L171

So, can we remove the name from the emitter message since we already have a name component?

Reply:
Also, we might need to handle name processing differently. If no name is specified by the user, we append the entity ID to the name, which should be unique. If the user defines a custom name, do we need to make sure this is still a unique name? Unique names are mentioned in the plugin documentation: https://github.com/ignitionrobotics/ign-gazebo/blob/9ed110b9105131a8c82ae86013bd8fdb4264fc21/src/systems/particle_emitter/ParticleEmitter.hh#L38-L39

Perhaps we should append the entity ID to the end of a user-defined name as well?

Reply:
I added `allow_renaming` plugin param to let user decide if renaming is allow or not - similar to the way it's done in the [breadcrumb system](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/systems/breadcrumbs/Breadcrumbs.cc#L74) 62646ea956ba5445bb5410327258c034139c6162

Reply:
Thanks :+1: I think the name field in the particle message may be removable just like the pose field we discussed. I'll leave this conversation marked as unresolved so that we remember to re-visit this once we have merged the PRs for parts 3 and 4.